hw_4_js/exercises.js

```diff
@@ -0,0 +1,415 @@
+  }
+}
+
+/**
+ * Exercise 12
+ *
+ * Write a function that returns the 2 smallest values of an array
+ * [4,3,2,1] => [1,2]
+ *
+ * Use: .sort()
+ */
+
+const get2SmallestValues = (numbers) => {
+  return numbers.sort().slice(0, 2)
+}
+
+/**
+ * Exercise 13
+ *
+ * Implement a function, at the input of which an object of the following form:
+ * {
+ * firstName: 'Peter',
+ * secondName: 'Vasiliev',
+ * patronymic: 'Ivanovich'
+ *}
+ * output line with the message 'Name: Petr Ivanovich Vasiliev'
+ */
+
+const getFullName = (user) => {
+  return 'Name: ' + user.firstName + ' ' + user.patronymic + ' ' + user.secondName
+}
+
+/**
+ * Exercise 14
+ *
+ * Implement a function that takes 2 arguments as input: an array of numbers and a multiplier,
+ * a returns an array of the original array, each element of which has been multiplied by a factor:
+ *
+ * [1,2,3,4], 5 => [5,10,15,20]
+ *
+ * Use: map
+ */
+
+const multiplyTo = (numbers, multiplier) => {
+  return numbers.map(n => n * multiplier)
+}
+
+/**
+ * Exercise 15
+ *
+ * Implement a function that takes 2 arguments as input: an array and a franchise,
+ * and returns a string with the names of the heroes separated by a comma:
+ *
+ *[
+ * {name: "Batman", franchise: "DC"},
+ * {name: "Ironman", franchise: "Marvel"},
+ * {name: "Thor", franchise: "Marvel"},
+ * {name: â€œSupermanâ€, franchise: â€œDCâ€}
+ *],
+ * Marvel
+ * => Ironman, Thor
+ *
+ * Use: filter, map, join
+ */
+
+const getCharacterNames = (characters, franchise) => {
+  return characters
+    .filter(c => c.franchise == franchise)
+    .map(c => c.name)
+    .join(', ')
+}
+
+// ----==== Advanced exercises (8 items) ====----
+/**
+ * Exercise 16
+ *
+ * Write a function that returns an array of the smallest row values of a two-dimensional array
+ *[
+ * [10,1,300,4],
+ * [20,2,300,400],
+ * [30,3,300,4],
+ * [40,4,300,4],
+ *]
+ * => [1,2,3,4]
+ */
+
+const getSmallestRow = (numbers) => {
+  return numbers.reduce((acc, current) => acc.concat([Math.min(...current)]), [])
+}
+
+/**
+ * Exercise 17
+ *
+ * Write a function that returns an array of the smallest column values of a two-dimensional array
+ *[
+ * [1,2,3,4],
+ * [1,2,3,4],
+ * [1,2,30,4],
+ * [1,2,3,40],
+ *]
+ * => [1,2,3,4]
+ *
+ * Here I used solution of previous problem, but for transposed matrix
+ */
+
+const getSmallestColumn = (numbers) => {
+  return numbers[0]
+    .map((_, columnNumber) => numbers.map(row => row[columnNumber]))
+    .reduce((acc, next) => acc.concat([Math.min(...next)]), [])
+}
+
+/**
+ * Exercise 18
+ *
+ * Write a function that returns the 2 biggest value of an array
+ * [4,3,2,1] => [4,3]
+ */
+
+const get2BiggestValues = (numbers) => {
+  const reducer = ([biggest1, biggest2], n) => {
+    if (biggest1 < n) {
+      biggest2 = biggest1
+      biggest1 = n
+    } else if (biggest2 < n) {
+      biggest2 = n
+    }
+    return [biggest1, biggest2]
+  }
+
+  return numbers.slice(2).reduce(reducer, [numbers[0], numbers[1]])
```

Comment:
Look at your implementation of `get2SmallestValues` :)
You can easily find how to take 2 biggest

Reply:
I thought that since this task in an advanced section it should be done in a more effective way than just sorting the array. Here time complexity is O(n) but sorting is O(n*logn) (worst case is O(n^2)) and I agree that in real life I would prefer to use sorting rather than this algo.